src/rulesets/Base/mapreduce.jl

```diff
@@ -67,30 +67,55 @@ function rrule(
 end
 
 function rrule(
-    config::RuleConfig{>:HasReverseMode}, ::typeof(sum), f, xs::AbstractArray; dims=:
-)
-    fx_and_pullbacks = map(x->rrule_via_ad(config, f, x), xs)
-    y = sum(first, fx_and_pullbacks; dims=dims)
+    config::RuleConfig{>:HasReverseMode}, ::typeof(sum), f::F, xs::AbstractArray; dims=:
+) where {F}
+    project = ProjectTo(xs)
 
-    pullbacks = last.(fx_and_pullbacks)
+    if _uses_input_only(f, eltype(xs))
+        # Then we can compute the forward pass as usual, save nothing but `xs`:
+        y = sum(f, xs; dims=dims)
+        function sum_pullback_easy(dy)
+            dxs = unthunk(dy) .* conj.(only.(only.(derivatives_given_output.(nothing, f, xs))))
+            return (NoTangent(), NoTangent(), project(dxs))
+        end
+        return y, sum_pullback_easy
+    end
 
-    project = ProjectTo(xs)
+    # In the general case, we need to save all the pullbacks:
+    fx_and_pullbacks = map(x -> rrule_via_ad(config, f, x), xs)
+    y = sum(first, fx_and_pullbacks; dims=dims)
+
+    function sum_pullback_f(dy)
+        # For arrays of arrays, we ought to protect the element against broadcasting:
+        dys = dims isa Colon ? Ref(unthunk(dy)) : unthunk(dy)
+        if Base.issingletontype(F)
+            # Then at least `f` has no gradient. Note that broadcasting here
+            # gets the shape right with or without `dims` keyword.
+            dxs = broadcast(fx_and_pullbacks, dys) do (_, back), dy1
+                unthunk(last(back(dy1)))
+            end
+            return (NoTangent(), NoTangent(), project(dxs))
 
-    function sum_pullback(ȳ)
-        call(f, x) = f(x)
-        # if dims is :, then need only left-handed only broadcast
-        broadcast_ȳ = dims isa Colon  ? (ȳ,) : ȳ
-        f̄_and_x̄s = call.(pullbacks, broadcast_ȳ)
-        # no point thunking as most of work is in f̄_and_x̄s which we need to compute for both
-        f̄ = if fieldcount(typeof(f)) === 0 # Then don't need to worry about derivative wrt f
-            NoTangent()
         else
-            sum(first, f̄_and_x̄s)
+            # Most general case. If `f` were stateful, we would need to reverse the order
+            # of iteration here, but since this funciton makes no guarantee, even the primal
+            # result is then ill-defined.
+            df_and_dxs = broadcast(fx_and_pullbacks, dys) do (_, back), dy1
+                map(unthunk, back(dy1))
```

Comment:
the code before has no `unthunk` here.
Instead it only `unthunk`s the `dx` part (and it only does that because when writtien project didn't support `InplacableThunk`s.
I think it is unneeded.

`sum` for sure should be able to accept thunks

```suggestion
                back(dy1)
```

Reply:
Yes, for `df` this is certainly not needed. 

For `dxs` I'm less sure. Is it OK to leave an array of thunks? Before this PR, that isn't what you get, so perhaps something is unthunking them?
```
julia> using ChainRulesCore

julia> const CFG = ChainRulesTestUtils.ADviaRuleConfig();

julia> y, bk = rrule(CFG, sum, sum, [[1,2], [3,4]]); y
10

julia> bk(1)[3]
2-element Vector{Vector{Float64}}:
 [1.0, 1.0]
 [1.0, 1.0]

(jl_b9KHB4) pkg> st ChainRules
Status `/private/var/folders/yq/4p2zwd614y59gszh7y9ypyhh0000gn/T/jl_b9KHB4/Project.toml`
  [082447d4] ChainRules v1.14.0
```
ProjectoTo does not like them right now, see https://github.com/JuliaDiff/ChainRules.jl/pull/529/files#r759742098 .